perfkitbenchmarker/providers/gcp/gcp_dpb_dataflow.py

```diff
@@ -152,4 +186,303 @@ def GetMetadata(self):
     basic_data = super(GcpDpbDataflow, self).GetMetadata()
     basic_data['dpb_dataflow_runner'] = FLAGS.dpb_dataflow_runner
     basic_data['dpb_dataflow_sdk'] = FLAGS.dpb_dataflow_sdk
+    basic_data['dpb_job_id'] = self.job_id
     return basic_data
+
+  def GetJobStatus(self):
+    cmd = util.GcloudCommand(self, 'dataflow', 'jobs', 'show', self.job_id)
+    cmd.flags = {
+        'project': self.project,
+        'format': 'json',
+    }
+
+  def GetStats(self):
```

Comment:
Do you need to actually refresh self.job_stats at some point in the lifetime of this class? If not, I would recomend rewriting this method as a [cached property](https://docs.python.org/3/library/functools.html#functools.cached_property).

E.g:
```python
@functools.cached_property
def job_stats(self):
  ...
```
Then update all the references

Reply:
Updated to cached property. However, that impacts the [reference in benchmark](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/pull/3703/files#diff-e1531d19e0eadfb8770d6c74873dec725e2eb5627da62dfa72e74dd635dddaefR184) as well. So every `dpb_service_instance` type should be aware of the `job_stats` property (vs `GetStats()` method) and ideally override it, is that a reasonable expectation/contract?

Reply:
While I do agree with your point, GetStats is not a standard dpb_service_instance method (i.e. is unique to the Dataflow implementation), so you if you leave it as-is you would face the same problem. This doesn't have any runtime consequences as the part of the benchmark code that access this property is guarded by a service_type check.

I know this is not pretty, but for me it's good enough if you leave a TODO in the benchmark code.

Reply:
Sounds good! Will leave the new `job_stats`